evm/src/lookup.rs

```diff
@@ -1,127 +1,248 @@
-use std::cmp::Ordering;
-
 use itertools::Itertools;
-use plonky2::field::extension::Extendable;
+use plonky2::field::batch_util::batch_add_inplace;
+use plonky2::field::extension::{Extendable, FieldExtension};
 use plonky2::field::packed::PackedField;
-use plonky2::field::types::{Field, PrimeField64};
+use plonky2::field::polynomial::PolynomialValues;
+use plonky2::field::types::Field;
 use plonky2::hash::hash_types::RichField;
+use plonky2::iop::ext_target::ExtensionTarget;
+use plonky2::iop::target::Target;
 use plonky2::plonk::circuit_builder::CircuitBuilder;
+use plonky2_util::ceil_div_usize;
 
 use crate::constraint_consumer::{ConstraintConsumer, RecursiveConstraintConsumer};
+use crate::stark::Stark;
 use crate::vars::{StarkEvaluationTargets, StarkEvaluationVars};
 
-pub(crate) fn eval_lookups<F: Field, P: PackedField<Scalar = F>, const COLS: usize>(
-    vars: StarkEvaluationVars<F, P, COLS>,
-    yield_constr: &mut ConstraintConsumer<P>,
-    col_permuted_input: usize,
-    col_permuted_table: usize,
-) {
-    let local_perm_input = vars.local_values[col_permuted_input];
-    let next_perm_table = vars.next_values[col_permuted_table];
-    let next_perm_input = vars.next_values[col_permuted_input];
-
-    // A "vertical" diff between the local and next permuted inputs.
-    let diff_input_prev = next_perm_input - local_perm_input;
-    // A "horizontal" diff between the next permuted input and permuted table value.
-    let diff_input_table = next_perm_input - next_perm_table;
-
-    yield_constr.constraint(diff_input_prev * diff_input_table);
-
-    // This is actually constraining the first row, as per the spec, since `diff_input_table`
-    // is a diff of the next row's values. In the context of `constraint_last_row`, the next
-    // row is the first row.
-    yield_constr.constraint_last_row(diff_input_table);
+pub struct Lookup {
+    /// Columns whose values should be contained in the lookup table.
+    /// These are the f_i(x) polynomials in the logUp paper.
+    pub(crate) columns: Vec<usize>,
+    /// Column containing the lookup table.
+    /// This is the t(x) polynomial in the paper.
+    pub(crate) table_column: usize,
+    /// Column containing the frequencies of `columns` in `table_column`.
+    /// This is the m(x) polynomial in the paper.
+    pub(crate) frequencies_column: usize,
```

Comment:
Something that popped up when discussing logUp with Daniel. 
Not sure if this is the right place, but we might want to check (or declare) the maximum number of overall trace entries subject to the lookup (i.e. the number of columns, times their length) to satisfy 
```
num_total_logUp_entries < char(F).
```
This will be a practically pointless check for the Goldilocks field, but not necessarily for our `M31`.